src/vmspawn/vmspawn.c

```diff
@@ -595,6 +598,32 @@ static int start_tpm(sd_bus *bus, const char *scope, const char *our_runtime_dir
         return 0;
 }
 
+static int finalize_root(char **ret) {
+        int r;
+        _cleanup_(dissected_image_unrefp) DissectedImage *image = NULL;
+        _cleanup_free_ char *root = NULL;
+
+        assert(ret);
+
+        r = dissect_image_file_and_warn(arg_image, NULL, NULL, NULL, 0, &image);
+        if (r < 0)
+                return log_error_errno(r, "Failed to dissect image: %m");
+
+        if (image->partitions[PARTITION_ROOT].found)
+                root = strjoin("root=PARTUUID=", SD_ID128_TO_UUID_STRING(image->partitions[PARTITION_ROOT].uuid));
+        else if (image->partitions[PARTITION_USR].found)
+                root = strjoin("mount.usr=PARTUUID=", SD_ID128_TO_UUID_STRING(image->partitions[PARTITION_USR].uuid));
+        else
+                return -ENOENT;
```

Comment:
we generally want that functions either log about all errors or about none. here you don't log on an error path, but the other error paths actually do.